File: pz-mod/42/media/lua/client/PZAgent/adapters/World.lua
--[[
PZAgent.Adapters.World -- world.inspect.

Invariant: this file only reads. It constructs no timed action, queues nothing,
equips nothing and moves nobody. That is not a stylistic preference -- it is the
reason world.inspect is in PZAgent.Protocol.READ_ONLY_ACTIONS, which lets it run
in OBSERVE mode and without arming. An inspect that walked the character to see
round a corner would be a mutating command wearing a read-only command's
permissions, so the absence of any call to Toolkit.construct or Toolkit.enqueue
below is load-bearing.

Because it only reads, its postcondition is unusual and worth stating plainly:
what a look proves is that a reading was taken, so the evidence *is* the
findings. There is no before/after pair to record, since nothing changed. Every
number in the document comes from an accessor that answered; a field the build
does not expose is absent rather than defaulted, because "light level 0" and
"this build does not report light" are different facts and only one of them
means the room is dark.

Everything is bounded twice: at most Toolkit.MAX_SQUARE_OBJECTS objects are read
from any one square, and at most World.MAX_OBJECTS across the whole reading. A
warehouse floor is exactly the input that would otherwise produce a document too
large to write, and truncation is declared in the evidence rather than hidden.
]]

PZAgent = PZAgent or {}
PZAgent.Adapters = PZAgent.Adapters or {}

-- Guarded the way Toolkit.lua guards them: the engine loads
-- media/lua/client/**.lua in an order this file does not control, so whichever
-- adapter file the game reads first is the one that creates the registry.
PZAgent.Adapters.ALL = PZAgent.Adapters.ALL or {}
PZAgent.Adapters.BY_NAME = PZAgent.Adapters.BY_NAME or {}

local World = {}
PZAgent.Adapters.World = World

--- Squares out from the centre. Two is a five-by-five block: enough to answer
--- "what is in this room with me", small enough that the reading stays a
--- glance rather than a survey.
World.MAX_RADIUS = 2

--- Objects across the whole reading, however many squares it covers. Roughly
--- three squares' worth of Toolkit.MAX_SQUARE_OBJECTS, which is what a document
--- the sidecar has to parse every tick can carry.
World.MAX_OBJECTS = 96

--- Engine symbol the reading stands on. The per-square accessors are not
--- declared here because they are methods on a Java object rather than globals:
--- PZAgent.CapabilityRuntime can only probe a global, and a method is checked
--- where it is called, by Toolkit.call.
local REQUIRED_SYMBOLS = { "getCell" }

--- Argument kinds, spelled out rather than read from
--- PZAgent.CommandDispatcher.ARG, because the dispatcher is a sibling file and
--- nothing here may dereference a module at load time. A typo costs a refused
--- registration at load rather than a loosened check at dispatch.
local ARG = {
  NUMBER = "number",
  REF = "ref",
}

-- ---------------------------------------------------------------------------
-- reading one object
-- ---------------------------------------------------------------------------

--- True when the engine's own class test says `object` is a `className`.
---
--- `instanceof` is the game's helper and is the only authority worth trusting
--- about a Java class. When the build does not expose it, the caller falls back
--- to classifying by the accessors an object answers to -- weaker, but wrong in
--- the safe direction: a door misread as a plain object is reported as a plain
--- object, never as a door whose state is known.
local function isA(object, className)
  if type(instanceof) ~= "function" then
    return nil
  end
  local ok, result = pcall(instanceof, object, className)
  if not ok then
    return nil
  end
  return result == true
end

World.isA = isA

--- The sprite name of an object, which is what identifies a tile to a human.
local function spriteName(object)
  local Toolkit = PZAgent.Adapters.Toolkit
  local ok, sprite = Toolkit.call(object, "getSprite")
  if not ok or sprite == nil then
    return nil
  end
  return Toolkit.readStringOf(sprite, { "getName" })
end

--- What kind of thing an object is, as far as this build will say.
---
--- Class test first, accessor shape second. `isBarricaded` alone is not enough
--- to call something a door, because a window answers it too; the pair that
--- separates them is a lock (doors) from a smash state (windows).
local function classify(object)
  local Toolkit = PZAgent.Adapters.Toolkit
  -- Build 42: IsoDoor, IsoWindow, IsoThumpable and IsoObject:getContainer().
  if isA(object, "IsoDoor") then
    return "door"
  end
  if isA(object, "IsoWindow") then
    return "window"
  end
  if Toolkit.method(object, "isLocked") ~= nil and Toolkit.method(object, "IsOpen") ~= nil then
    return "door"
  end
  if Toolkit.method(object, "isSmashed") ~= nil then
    return "window"
  end
  if Toolkit.method(object, "getContainer") ~= nil then
    return "container"
  end
  return "object"
end

World.classify = classify

--- The open/locked/barricaded state of a door or window.
---
--- Read for every object rather than only for the ones classify named, so a
--- build that spells IsoDoor differently still reports the state it does
--- expose. A field the object does not answer to is left out.
local function openingState(object, record)
  local Toolkit = PZAgent.Adapters.Toolkit
  -- Build 42: IsoDoor:IsOpen(), :isLocked(), :isBarricaded();
  -- IsoWindow:IsOpen(), :isSmashed(), :isBarricaded().
  record.open = Toolkit.readBooleanOf(object, { "IsOpen", "isOpen" })
  record.locked = Toolkit.readBooleanOf(object, { "isLocked", "isLockedByKey" })
  record.barricaded = Toolkit.readBooleanOf(object, { "isBarricaded" })
  record.smashed = Toolkit.readBooleanOf(object, { "isSmashed" })
  return record
end

--- Describe the container an object holds, without listing what is in it.
---
--- The count and the capacity are what a planner needs to decide whether to
--- look inside; the contents are container.inspect's job and would blow this
--- document's bound on the first shelf. The reference is minted here so the
--- follow-up command has something to name, and it is built by PZAgent.Refs so
--- the two sides agree on where the delimiters fall.
local function containerRecord(ctx, object, x, y, z, objectIndex, record)
  local Toolkit = PZAgent.Adapters.Toolkit
  local ok, container = Toolkit.call(object, "getContainer")
  if not ok or container == nil then
    return record
  end
  record.container_type = Toolkit.readStringOf(container, { "getType" })
  record.container_capacity = Toolkit.readNumberOf(container, { "getCapacity" })
  local items = Toolkit.containerItems(container)
  if items ~= nil then
    record.container_items = items.total
    record.container_truncated = items.truncated == true
  end
  local ref = PZAgent.Refs.worldContainer(ctx.session_id, x, y, z, objectIndex, 0)
  if ref ~= nil then
    record.container_ref = ref
  end
  return record
end

--- One object standing on a square.
local function objectRecord(ctx, object, x, y, z, objectIndex)
  local Toolkit = PZAgent.Adapters.Toolkit
  local record = {
    index = objectIndex,
    name = Toolkit.readStringOf(object, { "getObjectName", "getName" }),
    sprite = spriteName(object),
    kind = classify(object),
  }
  openingState(object, record)
  return containerRecord(ctx, object, x, y, z, objectIndex, record)
end

-- ---------------------------------------------------------------------------
-- reading one square
-- ---------------------------------------------------------------------------

--- The floor tile under a square, or nil when this build does not report one.
local function floorOf(square)
  local Toolkit = PZAgent.Adapters.Toolkit
  -- Build 42: IsoGridSquare:getFloor() -> IsoObject.
  local ok, floor = Toolkit.call(square, "getFloor")
  if not ok or floor == nil then
    return nil
  end
  return spriteName(floor) or Toolkit.readStringOf(floor, { "getObjectName" })
end

--- How lit a square is, on the engine's own scale.
---
--- Absent rather than zero when the accessor is missing: a caller deciding
--- whether the character needs a torch must be able to tell "it is dark" from
--- "nobody measured".
local function lightOf(square, playerIndex)
  local Toolkit = PZAgent.Adapters.Toolkit
  -- Build 42: IsoGridSquare:getLightLevel(playerIndex) -> 0..1.
  local ok, level = Toolkit.call(square, "getLightLevel", playerIndex or 0)
  if not ok or type(level) ~= "number" or level ~= level then
    return nil
  end
  return level
end

--- Everything on one square, bounded by both budgets.
---
--- `budget.objects` is spent across the whole reading, so a single crowded
--- square cannot starve the rest of the block. What was left out is counted, not
--- dropped in silence: a planner that acts on a partial reading has to know it
--- was partial.
local function squareRecord(ctx, x, y, z, budget)
  local Toolkit = PZAgent.Adapters.Toolkit
  local square, missing = Toolkit.gridSquare(x, y, z)
  if square == nil then
    return nil, missing
  end
  local record = {
    ref = PZAgent.Refs.buildSquare(ctx.session_id, x, y, z),
    x = x,
    y = y,
    z = z,
    floor = floorOf(square),
    light = lightOf(square, ctx.player_index),
    objects = {},
  }
  local ok, objects = Toolkit.call(square, "getObjects")
  if not ok or objects == nil then
    record.objects_readable = false
    return record
  end
  record.objects_readable = true
  local size = Toolkit.listSize(objects)
  if size == nil then
    record.objects_readable = false
    return record
  end
  record.object_count = size
  local scanned = math.min(size, Toolkit.MAX_SQUARE_OBJECTS, budget.objects)
  for index = 0, scanned - 1 do
    local object = Toolkit.listGet(objects, index)
    if object ~= nil then
      record.objects[#record.objects + 1] = objectRecord(ctx, object, x, y, z, index)
      budget.objects = budget.objects - 1
    end
  end
  record.truncated = size > scanned
  record.dropped = size - scanned
  return record
end

World.squareRecord = squareRecord

-- ---------------------------------------------------------------------------
-- world.inspect
-- ---------------------------------------------------------------------------

--- The square the reading is centred on.
---
--- With no reference it is the square the character is standing on, which is
--- the only centre an inspect can pick without being told. The floor comes from
--- the character's own z, so "look around me" never reads the storey below.
local function centreOf(ctx, ref)
  local Toolkit = PZAgent.Adapters.Toolkit
  local reasons = Toolkit.reasons()
  if ref ~= nil then
    local parsed, parseError = PZAgent.Refs.parseSquare(ref)
    if parsed == nil then
      return nil, reasons.INVALID_REF, parseError
    end
    return { x = parsed.x, y = parsed.y, z = parsed.z }
  end
  local snapshot = Toolkit.snapshot(ctx)
  if type(snapshot) ~= "table" or type(snapshot.position) ~= "table" then
    return nil, reasons.PRECONDITION_FAILED, "the character reports no position to look around from"
  end
  local position = snapshot.position
  return { x = math.floor(position.x), y = math.floor(position.y), z = position.z }
end

World.centreOf = centreOf

--- Look at a square and its neighbours.
---
--- Read-only and synchronous: there is nothing to wait for, so `start` finishes
--- the command itself and declares no poll.
local Inspect = {
  action = "world.inspect",
  required_symbols = REQUIRED_SYMBOLS,
  args = {
    -- Optional: a reading with no centre is a reading around the character.
    ref = { type = ARG.REF, kinds = { square = true } },
    radius = { type = ARG.NUMBER, integer = true, min = 0, max = World.MAX_RADIUS },
  },
}

function Inspect.start(ctx, args)
  local Toolkit = PZAgent.Adapters.Toolkit
  local reasons = Toolkit.reasons()
  local ready, symbolCode, symbolDetail = Toolkit.requireSymbols(REQUIRED_SYMBOLS)
  if ready == nil then
    return nil, symbolCode, symbolDetail
  end
  local centre, centreCode, centreDetail = centreOf(ctx, args.ref)
  if centre == nil then
    return nil, centreCode, centreDetail
  end
  local radius = args.radius or 1
  local budget = { objects = World.MAX_OBJECTS }
  local squares = {}
  local scanned, loaded, unavailable = 0, 0, nil
  for dx = -radius, radius do
    for dy = -radius, radius do
      scanned = scanned + 1
      local record, missing = squareRecord(ctx, centre.x + dx, centre.y + dy, centre.z, budget)
      if record ~= nil then
        loaded = loaded + 1
        squares[#squares + 1] = record
      elseif missing ~= "IsoCell.getGridSquare" then
        -- The cell itself could not be reached. That is a capability gap rather
        -- than an unloaded chunk, and it makes the whole reading meaningless.
        unavailable = missing
      end
    end
  end
  if unavailable ~= nil then
    return Toolkit.unavailable(unavailable)
  end
  if loaded == 0 then
    -- Not one square answered, so nothing was observed. Reporting an empty
    -- document as a successful look is precisely the fabricated success this
    -- project forbids.
    return nil, reasons.TARGET_NOT_LOADED, "no square around the requested centre is loaded"
  end
  local state = Toolkit.state(ctx)
  state.evidence = {
    centre = centre,
    radius = radius,
    squares = squares,
    squares_scanned = scanned,
    squares_loaded = loaded,
    objects_seen = World.MAX_OBJECTS - budget.objects,
    objects_truncated = budget.objects <= 0,
    observed_at_ms = ctx.now_ms,
  }
  return "done"
end

-- ---------------------------------------------------------------------------
-- registration
-- ---------------------------------------------------------------------------

--- Publish an adapter where the loader that feeds PZAgent.CommandDispatcher
--- looks for it.
---
--- Replacing by action name rather than always appending: the engine reads this
--- file once, but a reload during development would otherwise leave two copies
--- of the same adapter in ALL, and the dispatcher refuses a second registration
--- for an action it already serves.
local function publish(adapter)
  local all = PZAgent.Adapters.ALL
  for index = 1, #all do
    if all[index].action == adapter.action then
      all[index] = adapter
      PZAgent.Adapters.BY_NAME[adapter.action] = adapter
      return adapter
    end
  end
  all[#all + 1] = adapter
  PZAgent.Adapters.BY_NAME[adapter.action] = adapter
  return adapter
end

World.Inspect = publish(Inspect)

return World

File: packages/pz_agent_core/src/pz_agent_core/actions/adapters/world.py
"""``world.inspect``: reading the squares around the character.

A read-only action's postcondition is unusual enough to be worth stating
plainly: what a look proves is that a reading was *taken*, so the evidence is
the findings themselves. There is no before/after pair, because nothing changed
— which is precisely why the check below is that the *observation* describes the
squares the command asked about. A report that names no requested square is not
a look at the requested block; it is a look at nothing, and §4.11's "не
передавать огромный список sprites" is about what a reading contains, not about
whether one happened.

Nothing here moves the character. That absence is load-bearing rather than
incidental: ``world.inspect`` is in
:data:`~pz_agent_core.protocol.READ_ONLY_ACTIONS`, so it runs in OBSERVE mode
and without arming, and an inspect that walked round a corner to see better
would be a mutating command wearing a read-only command's permissions.

It also names no ``required_capability``, and that is deliberate. Every probe in
:mod:`pz_agent_core.capabilities.probes` resolves a *Lua* symbol on the user's
install; what a look needs is the observation's ``nearby`` tier, which the mod
either produced or did not. Declaring a probe over Java accessors the Lua
scanner can never see would report ``unsupported`` on a perfectly good install,
so the gate is the honest one: no ``nearby`` tier, no inspect.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import ClassVar, Final

from ...protocol import (
    ActionName,
    Command,
    JsonDict,
    NearbyObject,
    NearbyView,
    Observation,
    ReasonCode,
    RefError,
    RefKind,
    RiskClass,
    SquareRef,
)
from ..adapter import Evidence, PreconditionFailed
from .common import check_args, read_count, read_ref, require_nearby, square_of
from .movement import SQUARE_OBJECT_KIND

__all__ = [
    "DEFAULT_INSPECT_POLL_MS",
    "DEFAULT_INSPECT_RADIUS",
    "DEFAULT_INSPECT_TIMEOUT_MS",
    "MAX_INSPECT_RADIUS",
    "MAX_REPORTED_SQUARES",
    "WorldInspectAdapter",
]

#: Squares out from the centre. Two is a five-by-five block — enough to answer
#: "what is in this room with me", small enough that the reading stays a glance
#: rather than a survey. The mod applies the same ceiling
#: (``adapters/World.lua``); stating it twice is deliberate, because a radius
#: this side waves through is one the mod refuses.
MAX_INSPECT_RADIUS: Final = 2
DEFAULT_INSPECT_RADIUS: Final = 1

#: The largest block a reading can cover, and therefore the bound on the square
#: list that ends up in the evidence.
MAX_REPORTED_SQUARES: Final = (2 * MAX_INSPECT_RADIUS + 1) ** 2

#: A look is answered by the next observation, so the budget is short: waiting
#: longer would not make the mod describe more of the world.
DEFAULT_INSPECT_TIMEOUT_MS: Final = 5_000
DEFAULT_INSPECT_POLL_MS: Final = 100


@dataclass(frozen=True, slots=True)
class _InspectSpec:
    """The normalised centre and radius of one reading.

    Parsed by ``validate``, emitted by ``build_args`` and re-parsed by
    ``verify``, which sees the *prepared* command — one parser for all three, so
    the block that is checked is the block that was asked for.
    """

    x: int
    y: int
    z: int
    radius: int

    @classmethod
    def parse(cls, command: Command, observation: Observation) -> _InspectSpec:
        check_args(command, allowed=("ref", "radius"))
        radius = read_count(
            command,
            "radius",
            default=DEFAULT_INSPECT_RADIUS,
            minimum=0,
            maximum=MAX_INSPECT_RADIUS,
        )
        raw = command.args.get("ref")
        if raw is None:
            # No centre named is "look around me", which is the only centre a
            # reading can pick for itself. The floor comes from the character's
            # own z, so it never reads the storey below.
            x, y, z = square_of(observation.player.position)
            return cls(x=x, y=y, z=z, radius=radius)
        given = read_ref(command, "ref", kind=RefKind.SQUARE)
        try:
            parsed = SquareRef.parse(given)
        except RefError as exc:
            raise PreconditionFailed(
                f"ref is not a well-formed square reference: {given!r}",
                reason_code=ReasonCode.INVALID_REF,
            ) from exc
        return cls(x=parsed.x, y=parsed.y, z=parsed.z, radius=radius)

    def as_args(self, session_id: str) -> JsonDict:
        return {
            "ref": str(SquareRef(session_id=session_id, x=self.x, y=self.y, z=self.z)),
            "radius": self.radius,
        }

    def squares(self) -> tuple[tuple[int, int, int], ...]:
        """Every square the reading covers, in a stable order."""
        return tuple(
            (self.x + dx, self.y + dy, self.z)
            for dx in range(-self.radius, self.radius + 1)
            for dy in range(-self.radius, self.radius + 1)
        )


def _reported_square(nearby: NearbyView, x: int, y: int, z: int) -> NearbyObject | None:
    """The mod's description of one square, or None when it did not report it."""
    for candidate in nearby.objects:
        if candidate.kind != SQUARE_OBJECT_KIND or candidate.position is None:
            continue
        if square_of(candidate.position) == (x, y, z):
            return candidate
    return None


@dataclass(frozen=True, slots=True)
class WorldInspectAdapter:
    """``world.inspect``: describe the block of squares around a centre.

    An unloaded square is not a failure — a block at the edge of the loaded
    world always has some — but a block where *nothing* answered is. That is the
    same line the mod draws (``loaded == 0`` is ``TARGET_NOT_LOADED`` there),
    and it is the difference between a partial reading and no reading at all.
    """

    timeout_ms: int = DEFAULT_INSPECT_TIMEOUT_MS
    poll_interval_ms: int = DEFAULT_INSPECT_POLL_MS

    name: ClassVar[ActionName] = ActionName.WORLD_INSPECT
    risk: ClassVar[RiskClass] = RiskClass.P0
    required_capability: ClassVar[str | None] = None

    def validate(self, command: Command, observation: Observation) -> None:
        _InspectSpec.parse(command, observation)
        # Refuses with CAPABILITY_UNAVAILABLE when the mod produced no
        # surroundings: a look whose findings cannot be read back is a look
        # whose result could only ever be asserted.
        require_nearby(observation)

    def build_args(self, command: Command, observation: Observation) -> JsonDict:
        return _InspectSpec.parse(command, observation).as_args(command.session_id)

    def verify(self, command: Command, before: Observation, after: Observation) -> Evidence | None:
        spec = _InspectSpec.parse(command, before)
        if after.nearby is None:
            return None
        requested = spec.squares()
        described: list[JsonDict] = []
        for x, y, z in requested:
            square = _reported_square(after.nearby, x, y, z)
            if square is None:
                continue
            described.append(
                {
                    "ref": square.ref,
                    "x": x,
                    "y": y,
                    "z": z,
                    "semantics": sorted(square.semantics),
                }
            )
        if not described:
            return None
        return Evidence(
            kind="squares_described",
            observation_seq=after.seq,
            observed={
                "centre": {"x": spec.x, "y": spec.y, "z": spec.z},
                "radius": spec.radius,
                "squares_requested": len(requested),
                "squares_described": len(described),
                "squares_not_reported": len(requested) - len(described),
                "squares": described[:MAX_REPORTED_SQUARES],
            },
        )
